rpcs3/Emu/Cell/lv2/sys_game.cpp

```diff
@@ -5,52 +5,104 @@
 #include "Emu/System.h"
 #include "Emu/system_config.h"
 
-#include <thread>
+#include "Emu/IdManager.h"
+#include "Utilities/Thread.h"
 
 #include "sys_game.h"
 
 LOG_CHANNEL(sys_game);
 
-atomic_t<bool> watchdog_stopped = true;
-atomic_t<u64> watchdog_last_clear;
-u64 get_timestamp()
+struct watchdog_t
 {
-	return (get_system_time() - Emu.GetPauseTime());
-}
+	struct alignas(16) control_t
+	{
+		bool needs_restart = false;
+		bool active = false;
+		char pad[sizeof(u64) - sizeof(bool) * 2]{};
+		u64 timeout = umax;
+	};
 
-error_code _sys_game_watchdog_start(u32 timeout)
-{
-	sys_game.trace("sys_game_watchdog_start(timeout=%d)", timeout);
+	atomic_t<control_t> control{};
 
-	if (!watchdog_stopped)
+	void operator()()
 	{
-		return CELL_EABORT;
-	}
+		u64 start_time = get_system_time();
+		u64 old_time = start_time;
+		u64 current_time = old_time;
 
-	auto watchdog = [=]()
-	{
-		while (!watchdog_stopped)
+		constexpr u64 sleep_time = 50'000;
+
+		while (thread_ctrl::state() != thread_state::aborting)
 		{
-			if (Emu.IsStopped() || get_timestamp() - watchdog_last_clear > timeout * 1000000)
+			if (Emu.IsPaused())
 			{
-				watchdog_stopped = true;
-				if (!Emu.IsStopped())
+				start_time += current_time - old_time;
+				old_time = current_time;
+				thread_ctrl::wait_for(sleep_time);
+				current_time = get_system_time();
+				continue;
+			}
+
+			old_time = std::exchange(current_time, get_system_time());
+
+			auto old = control.fetch_op([&](control_t& data)
+			{
+				if (data.needs_restart)
 				{
-					sys_game.warning("Watchdog timeout! Restarting the game...");
-					Emu.CallFromMainThread([]()
-						{
-							Emu.Restart();
-						});
+					data.needs_restart = false;
+					return true;
 				}
-				break;
+
+				return false;
+			}).first;
+
+			if (old.active && old.needs_restart)
+			{
+				start_time = current_time;
+				old_time = current_time;
+				continue;
+			}
+
+			if (old.active && current_time - start_time >= old.timeout)
+			{
+				sys_game.warning("Watchdog timeout! Restarting the game...");
+
+				Emu.CallFromMainThread([]()
+				{
+					Emu.Restart();
+				});
+
+				return;
 			}
-			std::this_thread::sleep_for(1s);
+
+			thread_ctrl::wait_for(sleep_time);
 		}
-	};
+	}
+};
+
+error_code _sys_game_watchdog_start(u32 timeout)
```

Comment:
Could you help me confirm if the param 1 is indeed u32?
I was unable to check as in Ghidra it was shown as` undefined`, and the psdevwiki just said it's `int time(r)`.

Reply:
Yes I've checked,

Reply:
Would you mind sending me the disassembly of sys_game syscalls so that I might be able to implement them more accurately in the future? Thank you.

Reply:
IDA file. [lv2_446_+_bss_data.zip](https://github.com/RPCS3/rpcs3/files/9804018/lv2_446_%2B_bss_data.zip)

Reply:
Thank you.